src/server/serve_dir.rs

```diff
@@ -34,9 +34,23 @@ impl<State> Endpoint<State> for ServeDir {
                     log::info!("File not found: {:?}", dir);
                     return Response::new(StatusCode::NotFound);
                 }
-                Ok(file) => {
-                    // TODO: verify this is a sub-path of the original dir.
-                    File::open(file).await.unwrap() // TODO: remove unwrap
+                Ok(mut file_path) => {
+                    // Verify this is a sub-path of the original dir.
+                    let mut file_iter = (&mut file_path).iter();
+                    for lhs in &dir {
+                        if Some(lhs) != file_iter.next() {
+                            return Response::new(StatusCode::Forbidden);
+                        }
+                    }
```

Comment:
Is this the same as?
```rust
                    if !(&mut file_path).iter().eq(&dir) {
                        return Response::new(StatusCode::Forbidden);
                    }
```

Reply:
@pickfire not quite: this will never resolve to true since the iterators are of different length. I made sure to try `eq` before doing it this way. If `file_path` implemented `ExactSizeIterator` we could use `len` with `take` to make sure the lengths are the same. But alas.

Reply:
@yoshuawuyts How about `all`?

Reply:
Oh yeah that might work. We should try it out!

Reply:
We could use this instead. It compiles.
```rust
                    if !dir.iter().all(|lhs| Some(lhs) == file_iter.next()) {
                        return Response::new(StatusCode::Forbidden);
                    }
```